pageserver/src/page_service.rs

```diff
@@ -899,165 +995,383 @@ impl PageServerHandler {
+                    BatchState::Building(_) | BatchState::ReadMessagesEnded(_) => {}
+                    BatchState::ExecutorEnded => unreachable!("we only set this here"),
                 }
-                tokio::select! {
-                    biased;
-                    _ = self.cancel.cancelled() => {
-                        // We were requested to shut down.
-                        info!("shutdown request received in page handler");
-                        return Err(QueryError::Shutdown)
+                *guard = BatchState::ExecutorEnded;
+            });
+            let mut stop = false;
+            while !stop {
+                match batch_rx.changed().await {
+                    Ok(()) => {}
+                    Err(_) => {
+                        debug!("batch_rx observed disconnection of batcher");
                     }
-                    res = pgb.flush() => {
-                        res?;
+                };
+                let maybe_batch = {
+                    let borrow = batch_rx.borrow();
+                    let mut guard = borrow.lock().unwrap();
+                    match &mut *guard {
+                        BatchState::Building(maybe_batch) => maybe_batch.take(),
+                        BatchState::ReadMessagesEnded(maybe_batch) => {
+                            debug!("upstream dead");
+                            stop = true;
+                            maybe_batch.take()
+                        }
+                        BatchState::ExecutorEnded => {
+                            unreachable!("we break out of this loop after we set this state");
+                        }
                     }
+                };
+                let Some(batch) = maybe_batch else {
+                    break;
+                };
+                notify_batcher.notify_one();
+                debug!("processing batch");
+                self.pagesteam_handle_batched_message(pgb_writer, *batch, ctx)
+                    .await?;
+            }
+            Ok(())
+        };
+
+        //
+        // Execute the stages until they exit.
+        //
+        // We can either run the pipeline as concurrent futures or we can
+        // run it in separate tokio tasks.
+        //
+        // In any case, we need to be responsive to cancellation (self.cancel).
+        // The style chosen here is that cancellation must propagate through the
+        // pipeline: if any stage dies, the whole pipeline dies.
+        //
+        // If the client communicates intent to end the pagestream sub-protocol,
+        // the Reader stage shuts down the pipeline cleanly by sending a `None`
+        // through the pipeline, resulting in all stages exiting cleanly after
+        // the last response has been produced.
+        //
+        // Unclean pipeline shutdown is initiated by Reader or Executor returning
+        // a QueryError. This bubbles up to the caller, which will shut down the connection.
```

Comment:
I'm not sure I see this.

Let's say that `protocol_pipelining_mode=PageServiceProtocolPipeliningMode::Tasks` and the call to `pagesteam_handle_batched_message`.

`handle_pagerequests_pipelined` returns an error, but `read_message_task` and `batcher_task` are still running. They'll run until a message is received an realise that downstream is dead. For the `ConcurrentFutures` case, I think it works fine.

Reply:
> For the ConcurrentFutures case, I think it works fine.

Actually, I'm not sure about that. `join!` polls the futures from a lambda so a failure in one of the futures doesn't bail the function (I think). This is problematic for cases where we want to kill the pagestream connection on the server side (e.g. wrong shard).

Reply:
Hmm yes, correct. Will address this tomorrow.

Reply:
Done in recent pushes.